cmd/root/sweet_test.go

```diff
@@ -36,76 +80,351 @@ func fileToExercise(t *testing.T, fileName string) (exercise Exercise) {
-						t.Fatalf("%s: got no error, want error %s", tc.name, tc.wantErr)
-					} else if got.name != tc.want.name || got.text != tc.want.text {
-						m := fmt.Sprintf("%s: expected exercise don't match\n", tc.name)
-						m += fmt.Sprintf("got name  %s\n", got.name)
-						m += fmt.Sprintf("want name %s\n", tc.want.name)
-						m += fmt.Sprintf("name bytes  %v\n", []byte(got.name))
-						m += fmt.Sprintf("want bytes  %v\n\n", []byte(tc.want.name))
-						m += fmt.Sprintf("got  text %s\n", got.text)
-						m += fmt.Sprintf("want text %s\n", tc.want.text)
-						m += fmt.Sprintf("text bytes  %v\n", []byte(got.text))
-						m += fmt.Sprintf("want bytes  %v\n\n", []byte(tc.want.text))
-						t.Fatal(m)
-					}
-				} else {
-					if tc.wantErr == nil {
-						t.Fatalf("%s: got error %s, wanted no error ", tc.name, gotErr.Error())
-					}
+					t.Fatalf("%s wanted error, got nil\n", name)
+				}
+				if gotErr.Error() != wantErr.Error() {
+					t.Fatalf("%s wanted error msg \"%s\", got \"%s\"", name, wantErr.Error(), gotErr.Error())
 				}
 			},
+		},
+		{
+			// WARNING: This test might be flaky if the random
+			// selected exercise happens to be the go exercise.
+			args: []string{
+				"-l",
+				"go",
+			},
+			check: func(got Exercise, gotErr error) {
+				name := "random exercise, language selected"
+				want := testExercises[2]
+				if gotErr != nil {
+					t.Fatalf("%s wanted no error, got %s\n", name, gotErr)
+				}
+				if !got.matches(want) {
+					m := fmt.Sprintf("\n%s got\n", name)
+					m += got.details()
+					m += fmt.Sprintf("\nwanted\n")
+					m += want.details()
+					t.Fatal(m)
+				}
+			},
+		},
+		{
+			args: []string{
+				"-l",
+				"ts",
+			},
+			check: func(got Exercise, gotErr error) {
+				name := "random exercise, language selected, but not available"
+				wantErr := errors.New("failed to find exercise matching language ts")
+				if gotErr == nil {
+					t.Fatalf("%s wanted error, got nil\n", name)
+				}
+				if gotErr.Error() != wantErr.Error() {
+					t.Fatalf("%s wanted error msg \"%s\", got \"%s\"", name, wantErr.Error(), gotErr.Error())
+				}
+
+			},
+		},
+	}
+
+	for i, tc := range testCases {
+		cmd := mockCmd(tc)
+		if err := cmd.Execute(); err != nil {
+			t.Fatalf("mock command no. %d failed to run: %s", i, err)
 		}
+	}
+}
+
+func TestFromArgsWithStdin(t *testing.T) {
+	// Test environment setup
+	tmpExercisesDir := t.TempDir()
+	t.Setenv("SWEET_EXERCISES_DIR", tmpExercisesDir)
+
+	// Create a tmp file. This will replace os.Stdin
+	tmp, err := os.CreateTemp(".", "stdin")
+	if err != nil {
+		t.Error("Failed to create tmp file")
+	}
+	defer func() {
+		tmp.Close()
+		os.Remove(tmp.Name())
+	}()
+
+	wantText := "Hello from stdin!\n"
+	want := Exercise{
+		name: path.Join(".", tmp.Name()),
+		text: wantText,
+	}
+
+	// Write to tmp file
+	_, err = tmp.Write([]byte(wantText))
+	if err != nil {
+		t.Error("Failed to write to tmp file")
+	}
+	// Need to go back to the beginning to read Stdin
```

Comment:
~~Need to go back to the beginning to read Stdin~~ Go back to beginning of the file.